Custom cursors should not appear outside the web content viewport

hortont424 opened the pull request:
#### d606afcf01edb789f19727c895106feb7756f521
<pre>
Custom cursors should not appear outside the web content viewport
<a href="https://bugs.webkit.org/show_bug.cgi?id=294886">https://bugs.webkit.org/show_bug.cgi?id=294886</a>
<a href="https://rdar.apple.com/150695318">rdar://150695318</a>

Reviewed by Abrar Rahman Protyasha and Aditya Keerthi.

Align with the other browsers, and hide custom cursors when they are outside of the viewport.
This avoids the potential for confusion between the cursor and external UI.
Chrome and Firefox already behave this way.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::selectCursor):
Restrict custom cursor presentation such that they only appear when fully insid the viewport.

* LayoutTests/fast/events/mouse-cursor-hidden-when-out-of-bounds-expected.txt: Added.
* LayoutTests/fast/events/mouse-cursor-hidden-when-out-of-bounds.html: Added.
Add a new test that verifies this behavior.

* LayoutTests/fast/events/mouse-cursor.html:
Adjust an existing test that didn&apos;t expect this behavior, by adding a margin.

* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-site-isolation/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
Propagate test expectations from mouse-cursor.html, which the new test was copied from.

Canonical link: <a href="https://commits.webkit.org/296590@main">https://commits.webkit.org/296590@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7aa2cb067862aa3cabb9181219b2b35296f790a4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108964 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28624 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19049 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/114173 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59302 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110927 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29308 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37192 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82802 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111912 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23303 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98152 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63241 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22722 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16287 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58870 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92678 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16335 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117293 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36014 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26619 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91814 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36386 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94414 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91622 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23326 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36531 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14291 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/31880 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35912 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41437 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35613 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38953 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37297 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on LayoutTests/fast/events/mouse-cursor-hidden-when-out-of-bounds.html:
```diff
@@ -0,0 +1,97 @@
+<!DOCTYPE html>
+<html>
+<head>
+<script src="../../resources/js-test-pre.js"></script>
```
```suggestion
<script src="../../resources/js-test.js"></script>
```

And no need for the -post.js below.

hortont424 committed:
Custom cursors should not appear outside the web content viewport
https://bugs.webkit.org/show_bug.cgi?id=294886
rdar://150695318

Reviewed by Abrar Rahman Protyasha and Aditya Keerthi.

Align with the other browsers, and hide custom cursors when they are outside of the viewport.
This avoids the potential for confusion between the cursor and external UI.
Chrome and Firefox already behave this way.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::selectCursor):
Restrict custom cursor presentation such that they only appear when fully insid the viewport.

* LayoutTests/fast/events/mouse-cursor-hidden-when-out-of-bounds-expected.txt: Added.
* LayoutTests/fast/events/mouse-cursor-hidden-when-out-of-bounds.html: Added.
Add a new test that verifies this behavior.

* LayoutTests/fast/events/mouse-cursor.html:
Adjust an existing test that didn't expect this behavior, by adding a margin.

* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-site-isolation/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
Propagate test expectations from mouse-cursor.html, which the new test was copied from.

Canonical link: https://commits.webkit.org/296590@main
